crypto/params.c

```diff
@@ -0,0 +1,375 @@
+        switch (p->buffer_size) {
+        CASE(uint32_t);
+        CASE(uint64_t);
+        }
+    }
+    return 0;
+#undef CASE
+}
+
+int OSSL_PARAM_set_double(const OSSL_PARAM *p, double val)
+{
+#define CASE(type) \
+    case sizeof(type): \
+        *(type *)p->buffer = (type)val; \
+        SET_RETURN_SIZE(p, sizeof(type)); \
+        return 1
+
+    if (p == NULL)
+        return 0;
+
+    SET_RETURN_SIZE(p, sizeof(double));
+    switch (p->data_type) {
+    case OSSL_PARAM_REAL:
+        switch (p->buffer_size) {
+        CASE(double);
+        }
+        break;
+    case OSSL_PARAM_INTEGER:
+        switch (p->buffer_size) {
+        CASE(int32_t);
+        CASE(int64_t);
+        }
+        break;
+    case OSSL_PARAM_UNSIGNED_INTEGER:
+        switch (p->buffer_size) {
+        CASE(uint32_t);
+        CASE(uint64_t);
+        }
+        break;
+    }
+    return 0;    
+#undef CASE
+}
+
+OSSL_PARAM OSSL_PARAM_construct_utf8_string(const char *key, char *buf,
+                                            size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_UTF8_STRING, buf, bsize,
+                                rsize);
+}
+
+int OSSL_PARAM_get_utf8_string(const OSSL_PARAM *p, char **val, size_t max_len)
+{
+    char *q;
+
+    if (val == NULL || p == NULL || p->data_type != OSSL_PARAM_UTF8_STRING)
+        return 0;
+
+    q = *val;
+    if (q == NULL) {
+        if ((q = OPENSSL_malloc(p->buffer_size)) == NULL)
+            return 0;
+        *val = q;
+        memcpy(q, p->buffer, p->buffer_size);
+        return 1;
+    }
+    if (max_len < p->buffer_size)
+        return 0;
+    memcpy(q, p->buffer, p->buffer_size);
+    return 1;
+}
+
+int OSSL_PARAM_set_utf8_string(OSSL_PARAM *p, char *val)
+{
+    if (p == NULL)
+        return 0;
+
+    SET_RETURN_SIZE(p, 0);
+    if (val != NULL) {
+        const size_t len = strlen(val) + 1;
+
+        SET_RETURN_SIZE(p, len);
+        if (p->data_type == OSSL_PARAM_UTF8_STRING && p->buffer_size >= len) {
+            memcpy(p->buffer, val, len);
+            return 1;
+        }
+    }
+    return 0;
+}
+
+OSSL_PARAM OSSL_PARAM_construct_octet_string(const char *key, void *buf,
+                                             size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_OCTET_STRING, buf, bsize,
+                                rsize);
+}
+
+int OSSL_PARAM_get_octet_string(const OSSL_PARAM *p, void **val, size_t max_len,
+                                size_t *used)
+{
+    size_t len;
+    char *q;
+
+    if (val == NULL || p == NULL || p->data_type != OSSL_PARAM_OCTET_STRING)
+        return 0;
+
+    *used = len = p->buffer_size;
+    if (*val == NULL) {
+        if ((q = OPENSSL_malloc(len)) == NULL)
+            return 0;
+        *val = q;
+        memcpy(q, p->buffer, len);
+        return 1;
+    }
+    if (max_len < len)
+        return 0;
+    memcpy(*val, p->buffer, len);
+    return 1;
+}
```

Comment:
Have you noticed that UTF8_STRING and OCTET_STRING are really exactly the same, apart from the data type?
You could actually use the exact same function for both.  Or, you could have an inner function that does what your octet_string functions do, and then just:

``` C
int OSSL_PARAM_get_utf8_string(const OSSL_PARAM *p, void **val, size_t max_len)
{
    size_t unused = 0;
    return get_string_int(p, val, max_len, &unused);
}

int OSSL_PARAM_get_octet_string(const OSSL_PARAM *p, void **val, size_t max_len,
                                size_t *used)
{
    return get_string_int(p, val, max_len, used);
}
```

Likewise for the `_set_` variants

Reply:
They shouldn't be identical but they are close enough to merge together.  The length of a utf8 string can be determined using strlen, but not for an octet string.

Reply:
Hence `size_t unused` in my utf8_string wrapper...

Reply:
I've refactored all of these into common functions as far as is possible.